core/corehttp/hostname.go

```diff
@@ -334,11 +390,16 @@ func toSubdomainURL(hostname, path string, r *http.Request) (redirURL string, ok
 		// if object turns out to be a valid CID,
 		// ensure text representation used in subdomain is CIDv1 in Base32
 		// https://github.com/ipfs/in-web-browsers/issues/89
-		rootID, err = cid.NewCidV1(multicodec, rootCid.Hash()).StringOfBase(mbase.Base32)
+		rootCID = cid.NewCidV1(multicodec, rootCID.Hash())
+		rootID, err = rootCID.StringOfBase(mbase.Base32)
```

Comment:
Note: we still default to Base32 here..

Reply:
Is the idea that even if we changed the default libp2p-key output to be base36 that people who put base32 RSA libp2p-keys into the browser domain or path wouldn't end up with the domain (and therefore origin) changed?

Similarly, is this solution too flexible? Do we need a canonical base encoding for the domains so that data stored in browser local storage isn't lost?

Reply:
The idea here is to PR a surgical change to **use Base36 only where it is really needed and continue with Base32 as default for subdomains** (so already existing subdomain Origins don't change, and we don't need to worry about any data lost).

I don't believe we need to add any additional normalization.

Reply:
My question was basically about what we want to happen if I use base36 where I could've used base32.

Leaving as is may lead to me having two different views (with different local data) for the same website.

Making them canonical is both a bit of a pain and uses an arbitrary rule that we likely end up stuck with (like the Cidv0-1 size cutoff).

Wanted to double check we are ok with the pros and cons of our solution not

Reply:
My question was basically about what we want to happen if I use base36 where I could've used base32.

Leaving as is may lead to me having two different views (with different local data) for the same website.

Making them canonical is both a bit of a pain and uses an arbitrary rule that we likely end up stuck with (like the Cidv0-1 size cutoff).

Wanted to double check we are ok with the pros and cons of our solution not

Reply:
I think RSA in Base32 is ok for now (when RSA is the default and CLI output is in Base58btc), use of Base36 is still non-existent. 

My thinking was that when we switch libp2p-key in CLI output to CIDv1 as part of ed25519 epic (https://github.com/ipfs/go-ipfs/issues/6916), then we will make a decision about Base36 being "canonical" base for `/ipns/` and then may want to adjust this section to use the same Base36 for all CID libp2p-keys (RSA and ED25519)

But as long RSA is the default and Base36 is fairly new, I'd rather not rock the boat prematurely.

Reply:
> Yes, the idea here is to PR a surgical change to use Base36 only where it is really needed and continue with Base32 as default for subdomains (so already existing subdomain Origins don't change, and we don't need to worry about any data lost).

I really think this is the wrong tradeoff. The amount of "data lost" right now is still absolutely minimal, compared to what is in store if we end up needing to switch to b36.

@aschmahmann is correct: if you do not bite the bullet ans make the switch now, you are stuck with yet another special case forever†

† forever, n: an excruciatingly long period time

Reply:
@ribasushi @lidel I suspect for IPNS if we switched to base36 everywhere probably no one would mind since those users are probably keeping updated so that they can get IPNS performance improvements as they come. However, how bad is the damage for IPFS users in subdomains? We pushed that out in 0.5 so it's still pretty new, but there are presumably some users already relying on it.

It seems like we might be able to get away with just b36 encoding all the subdomains to avoid worrying about:
1) Canonical encodings: What happens if a user uses a b36 subdomain for something that could have been b32? Won't it be weird that the state of those two websites could be different?
2) Annoying rules: We can solve 1, by just coming up with some canonicalization like "it's b32 unless it won't fit then it's b36", but that may be annoying to think about longer term compared with just deciding to use b36 everywhere.

Reply:
Ad. Canonical encodings: IMO not an issue in practice. The same HTML can be loaded from different hostnames on the web, and each gets distinct Origin. I think its a feature of IPFS to be able to represent the same CID with different base, we should not force people to use b36 if their CID works with b32 and they prefer it for some reason and explicitly used that in subdomain (eg. third party tooling implements RFC version of b32, and no b36 yet).

----

Just to be clear: I am all for simplicity of using b36 for entire `/ipns/` (including RSA).

My only concern was that **we've just added b36 in the last release** and if we flip the switch and use it for all libp2p-keys on subdomains, and people start copying and pasting those CIDs, various things will break for people: mind that there is no support for b36 in non-go non-js IPFS-related libs and clients at all etc, and we have no bandwidth in Q3 to help with changing that.

That is why I wanted to keep status quo for at least one or two more releases, and effectively only use b36 for ed25519 here.

That being said, 
@ribasushi @aschmahmann if you both are ok with using b36 for entire `/ipns/` (including RSA), I can update this PR to do just that – let me know.

Reply:
Formal :+1: from me on flipping **as much as possible** to `b36`. I fully recognize the pain it **will** cause. That pain is utterly dwarfed by the cost of doing another switch further down the road.

Reply:
I'd like to be as consistent as possible. Given that IPNS isn't widely used for obvious reasons anyways, I'd suggest at least using base36 for IPNS. I'm fine not canonicalizing CIDs.

> The same HTML can be loaded from different hostnames on the web, and each gets distinct Origin.

That's fine. The problem is if we decide to start canonicalizing CIDs to base36 for consistency in the future. If we do that, we'll be _changing_ the origin.

Reply:
:+1: I tend to think that we might as well do the b36 encoding for IPNS. IPNS users tend to be pretty happy when fixes show up and there'll probably be at least a few fixes in 0.7 or 0.8 that make it worth their while.

As an aside I've been thinking about this a bit and I suspect the confusion here stems from us not really telling people what the "correct" way is for users to describe an IPFS/IPNS URL. Do we even have a single answer here, or is it context dependent? For example, we'd like people to ideally use `ipfs://b32encodedID`, but since `ipfs://` isn't yet widely supported we probably like `dweb.link/ipfs/b32encodedID`. Not having an canonical answer means we're stuck guessing about how people "could" be using URLs and so makes reasoning about upgrade paths difficult.

Reply:
Great! I'm updated this PR to use b35 for ALL libp2p-keys in subdomains (79d571f272d640f0d4164f6f9d94c5fb44983822)

@aschmahmann I believe that was the last thing blocking this PR – mind taking a final look and merging if it is ok?

ps. regarding canonical address, for the time being it is: `https://dweb.link/ipfs/{cid}` | `dweb.link/ipns/{libp2p-key}`
Works everywhere, can be shared and loads fine, and is future proof. 
We need to work with gui/docs/ecosystem to make that more evident in our apps, docs and comms.

Reply:
Thanks will take a look. Is there any base encoding recommended for the dweb.link paths, or is that left unspecified? 💯 on documenting the path based recommended addressing though since it will at least allow us to point people at something if they ever run into issues copy-pasting a subdomain-based URL.